llvm/include/llvm/CodeGen/SDPatternMatch.h

```diff
@@ -547,6 +547,39 @@ struct BinaryOpc_match {
   }
 };
 
+/// Matching while capturing mask
+template <typename T0, typename T1, typename T2> struct SDShuffle_match {
+  T0 Op1;
+  T1 Op2;
+  T2 Mask;
+
+  SDShuffle_match(const T0 &Op1, const T1 &Op2, const T2 &Mask)
+      : Op1(Op1), Op2(Op2), Mask(Mask) {}
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    if (auto *I = dyn_cast<ShuffleVectorSDNode>(N)) {
+      return Op1.match(Ctx, I->getOperand(0)) &&
+             Op2.match(Ctx, I->getOperand(1)) && Mask.match(I->getMask());
+    }
+    return false;
+  }
+};
+struct m_Mask {
+  ArrayRef<int> MaskRef;
```

Comment:
This one does need to be a reference. The assignment in `match` needs to modify the ArrayRef<int> that the caller passed in.

Reply:
I am still not understanding passing an ArrayRef by reference. To clarify, I beleive what you and @mshockwave are suggesting is to implement `m_Mask` by reference and `m_SpecificMask` by value. 

The argument passed to `m_Mask` will be reference to an `ArrayRef<int>`. 

The argument passed to `m_SpecificMask` will be an `ArrayRef<int>`, copying the data. The data being `const [T] *Data` and `size_type Length`. 

- - -
Where I am confused,assuming what I stated above is correct: 

- In the IR, patternmatch.h, both `m_Mask` and `m_SpecificMask` are implement as pass-by-reference (as is `m_SplatOrPoisonMask`, `m_ZeroMask` is by value). Is what we want to do different than what happens in the IR?

- I think I am not aware enough of the uses of matching in this context. Because of this, passing an ArrayRef by reference or by value seems almost the same, depending on the usecase. In both cases you have access to the underlying array, so modifcations can be made. Are we trying to avoid a copy of the `Length` field? Is there a case where we want to modify the `ArrayRef` object itself, but not via the underlying array? 

- - - 
Some clarification would be greatly appreciated, thank you!

Reply:
> In the IR, patternmatch.h, both m_Mask and m_SpecificMask are implement as pass-by-reference (as is m_SplatOrPoisonMask, m_ZeroMask is by value)

IR PatternMatch's m_SpecificMask is wrong (though it doesn't affect the correctness) because if you look info it, [it never assigns](https://github.com/llvm/llvm-project/blob/af83093933ca73bc82c33130f8bda9f1ae54aae2/llvm/include/llvm/IR/PatternMatch.h#L1849) value into MaskRef.

>  passing an ArrayRef by reference or by value seems almost the same,

It's functionally the same for both SDPatternMatch's and IR PatternMatch's m_SpecificMatch (though from a software engineering point of view, you should pass by value for m_SpecificMatch), but it's not the same for m_Mask: for m_Mask, you want to capture the mask value so you _have_ to use a reference (or pointer) to carry the result.

> In both cases you have access to the underlying array, so modifcations can be made

No, ArrayRef is [designed](https://github.com/llvm/llvm-project/blob/65a2eb0b1589590ae78cc1e5f05cd004b3b3bec5/llvm/include/llvm/ADT/ArrayRef.h#L29) to be a reference to an _immutable_ chunk of memory -- otherwise `ArrayRef::operator[]` should have a return type of `T&` instead of `const T&` --  so you cannot modify the underlying array. If you want to do that, there is `llvm::MutableArrayRef`.

Reply:
> * Are we trying to avoid a copy of the `Length` field

slightly tangent to this issue: the reason ArrayRef is trivially copyable, or, why we prefer copy by value is that ArrayRef only has two fields, and modern processors + ABI almost certain will pass these two fields in two registers. If we pass by reference, I think the compiler has to put the ArrayRef object on memory because we can only take address from lvalue. Once that happens, it's slower than passing by two registers.

Reply:
Thank you for the help. My most recent commit incorporates these changes. I believe my test logic is correct.

Uninitialzed masks:
```
  const std::array<int, 4> MaskData = {2, 0, 3, 1};      //used to construct vectorshuffle with `DAG->getVectorShuffle()` 
  const std::array<int, 4> OtherMaskData = {1, 2, 3, 4};
  ArrayRef<int> Mask;                                    //captured with m_mask in test
```

however to m_SpecificMask should we be passing `MaskData` or `Mask` by value for a comparison?

Reply:
> however to m_SpecificMask should we be passing MaskData or Mask by value for a comparison?

In this case you probably have to use `MaskData` because you haven't checked the correctness of `Mask` at the time you test m_SpecificMask. Unless you move that correctness check (line 145 to 147) to the front, in which case it doesn't matter which one you use.